docs/docs/specs/v0/composer.md

```diff
@@ -0,0 +1,190 @@
+
+Two levels of complexity are proposed in sequence:
+
+**Static chrome**, where we constrain the application chrome to be statically defined and not extensible by plugins. Elements of the chrome may each expose some API surface for plugins to extend, but the overall elements of the chrome are fixed and can be assumed by all plugins without explicit dependency expressions.
+
+**Dynamic chrome**, where the chrome and it's organization is also extensible and now plugins have to express explicitly the elements of chrome they depend on because their presence can no longer be assumed.
+
+These are in addition to another orthogonal dimension of complexity, the way we load plugins:
+
+**Static plugins**, where a `<Composer />` element is given a list of `Plugin[]` directly as props and has to be compiled into the app at build time.
+
+**Dynamic plugins**, where the app knows how to load plugins via ESM `import` calls in the browser at runtime.
+
+### 1. Static chrome (v1)
+
+Large screens are split vertically into `sidebar` and `content` areas.
+
+```
+.------------.
+|    |     o |
+|    |       |
+|    |       |
+|    |       |
+'------------'
+```
+
+Small screens turn the `sidebar` into a slide-over element that hides by default.
+
+The HALO button (o) is in one of the corners and can be used to access the DXOS Shell.
+
+The sidebar presents a `tree` of nodes which can be populated by plugins.
+
+The `tree` is a logical first solution to the problem of organization and represents a model users and developers are very familiar with and have grown to expect, having found the same model in almost all nearby and competitive products: Notion, Quip, Dropbox Paper, Google Docs, VSCode, Obsidian, ... etc.
+
+Plugins provide components that can fill the content area fully, one at a time, and can sense the state of the tree view (selection, nodes, etc).
+
+Search happens via an overlay [Cmd+K experience](https://cmdk.paco.me/) where a box pops over all the surfaces and presents a fuzzy-search with autocomplete which can be used to locate objects to navigate to, or actions and commands to invoke.
+
+#### Plugin interfaces
+
+```ts
+type MaybePromise<T> = T | Promise<T>;
+
+type Plugin = {
+  meta: {
+    // serializable, could be stored in a space.
+    id: string;
+    name: string;
+    description: string;
+  };
+  provides: {
+    tree: {
+      getTreeNodes(parent: TreeNode): Observable<TreeNode[]>;
+      getTreeNodeActions(parent: TreeNode): Observable<Action[]>;
+    };
+    content: {
+      getComponent(selection: TreeNode[]): MaybePromise<React.FC>;
+    };
+  };
+};
+
+type Effect = (state: ComposerState) => MaybePromise<ComposerState>;
+
+type Action = {
+  id: string;
+  label: string;
+  icon?: React.FC;
+  invoke(state: ComposerState): MaybePromise<Effect | Effect[]>;
+};
+
+type TreeNode<T = any> = {
```

Comment:
Helpful to think consider layers. E.g., "Action" seems out of context with "Tree" (i.e., horizontal concept vs. vertical concept).

Similarly I suggest removing "Composer" from framework type definitions.

Reply:
`ComposerState` renamed `AppState`

re: `Action` let's discuss, it's unclear the meaning of horizontal and vertical here.